pubsub/pulsar/pulsar.go

```diff
@@ -318,22 +334,73 @@ func parsePublishMetadata(req *pubsub.PublishRequest, schema schemaMetadata) (
 		msg.Value = obj
 	}
 
-	if val, ok := req.Metadata[deliverAt]; ok {
-		msg.DeliverAt, err = time.Parse(time.RFC3339, val)
-		if err != nil {
-			return nil, err
-		}
-	}
-	if val, ok := req.Metadata[deliverAfter]; ok {
-		msg.DeliverAfter, err = time.ParseDuration(val)
-		if err != nil {
-			return nil, err
+	for name, value := range req.Metadata {
+		switch name {
+		case partitionKey:
+			msg.Key = value
+		case deliverAt:
+			msg.DeliverAt, err = time.Parse(time.RFC3339, value)
+			if err != nil {
+				return nil, err
+			}
+		case deliverAfter:
+			msg.DeliverAfter, err = time.ParseDuration(value)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			if msg.Properties == nil {
+				msg.Properties = make(map[string]string)
+			}
+			msg.Properties[name] = value
 		}
 	}
 
 	return msg, nil
 }
 
+// default: shared
+func getSubscribeType(req pubsub.SubscribeRequest) pulsar.SubscriptionType {
+	metadata := req.Metadata
+
+	var subsType pulsar.SubscriptionType
+
+	subsTypeStr, _ := metadata[subscribeTypeKey]
```

Comment:
Should we lowercase this?

```suggestion
	subsTypeStr := strings.ToLower(metadata[subscribeTypeKey])
```